### PR TITLE
Fix Unordered Repeated Records

### PR DESCRIPTION
**Pull request by idreeskhan**
#219 and #217 

* Fixes comparisons between nested repeated records that contain different lengths
* Fixes logic to sort and compare nested records

### CODE REVIEW

File: ratatool-diffy/src/main/scala/com/spotify/ratatool/diffy/ProtoBufDiffy.scala [hunk 1]
```diff
@@ -59,40 +61,52 @@ extends Diffy[T](ignore, unordered, unorderedFieldKeys) {
...
-          unorderedFieldKeys.get(fullName).map(s => getField(getFieldDescriptor(f, s)) _)
-        val a = sortList(x.getField(f).asInstanceOf[java.util.List[AbstractMessage]], getFieldFn)
-        val b = sortList(y.getField(f).asInstanceOf[java.util.List[AbstractMessage]], getFieldFn)
-        if (f.getJavaType == JavaType.MESSAGE && unordered.exists(_.startsWith(s"$fullName."))
-            && unorderedFieldKeys.contains(fullName)) {
-          a.asInstanceOf[java.util.List[AbstractMessage]].asScala.zip(
-            b.asInstanceOf[java.util.List[AbstractMessage]].asScala).flatMap {
-              case (l, r) => diff(l, r, f.getMessageType.getFields.asScala, fullName)}
+        if (f.getJavaType == JavaType.MESSAGE
+          && unordered.contains(fullName)
```
Comment (anne-decusatis) [inline ratatool-diffy/src/main/scala/com/spotify/ratatool/diffy/ProtoBufDiffy.scala]: this contains check is already checked above

File: ratatool-diffy/src/main/scala/com/spotify/ratatool/diffy/ProtoBufDiffy.scala [hunk 2]
```diff
@@ -59,40 +61,52 @@ extends Diffy[T](ignore, unordered, unorderedFieldKeys) {
...
         }
       } else {
         f.getJavaType match {
           case JavaType.MESSAGE if !f.isRepeated =>
-            val a = getField(f)(x).asInstanceOf[AbstractMessage]
-            val b = getField(f)(y).asInstanceOf[AbstractMessage]
-            if (a == null && b == null) {
+            val a = x.flatMap(m => getField(f)(m).asInstanceOf[Option[AbstractMessage]])
+            val b = y.flatMap(m => getField(f)(m).asInstanceOf[Option[AbstractMessage]])
+            if (a.isEmpty && b.isEmpty) {
```
Comment (anne-decusatis) [inline ratatool-diffy/src/main/scala/com/spotify/ratatool/diffy/ProtoBufDiffy.scala]: I liked the case statement you changed the other file to use in the same part of that logic a bit better (but this is still ok)

File: ratatool-diffy/src/main/scala/com/spotify/ratatool/diffy/ProtoBufDiffy.scala [hunk 3]
```diff
@@ -59,40 +61,52 @@ extends Diffy[T](ignore, unordered, unorderedFieldKeys) {
...
-        val b = sortList(y.getField(f).asInstanceOf[java.util.List[AbstractMessage]], getFieldFn)
-        if (f.getJavaType == JavaType.MESSAGE && unordered.exists(_.startsWith(s"$fullName."))
-            && unorderedFieldKeys.contains(fullName)) {
-          a.asInstanceOf[java.util.List[AbstractMessage]].asScala.zip(
-            b.asInstanceOf[java.util.List[AbstractMessage]].asScala).flatMap {
-              case (l, r) => diff(l, r, f.getMessageType.getFields.asScala, fullName)}
+        if (f.getJavaType == JavaType.MESSAGE
+          && unordered.contains(fullName)
+          && unorderedFieldKeys.contains(fullName)) {
+          val l = x.flatMap(r =>
```
Comment (anne-decusatis) [inline ratatool-diffy/src/main/scala/com/spotify/ratatool/diffy/ProtoBufDiffy.scala]: similar comments in this as in above

File: ratatool-diffy/src/main/scala/com/spotify/ratatool/diffy/TableRowDiffy.scala [hunk 4]
```diff
@@ -47,48 +48,54 @@ class TableRowDiffy(tableSchema: TableSchema,
...
-            diff(l, r, f.getFields.asScala, fullName)
-          }
-        }
-        else {
-          val a = sortList(x.get(name).asInstanceOf[java.util.List[AnyRef]])
-          val b = sortList(y.get(name).asInstanceOf[java.util.List[AnyRef]])
-          if (a == b) Nil else Seq(Delta(fullName, Option(a), Option(b), delta(a, b)))
+        if (f.getType == "RECORD"
+          && unorderedFieldKeys.contains(fullName)
+          && unordered.contains(fullName)) {
```
Comment (anne-decusatis) [inline ratatool-diffy/src/main/scala/com/spotify/ratatool/diffy/TableRowDiffy.scala]: this is also checked above already; similar comments in this file to other files

File: ratatool-diffy/src/main/scala/com/spotify/ratatool/diffy/AvroDiffy.scala [hunk 5]
```diff
@@ -20,67 +20,79 @@ package com.spotify.ratatool.diffy
...
+              Option(r.get(name).asInstanceOf[java.util.List[GenericRecord]].asScala.toList))
+              .getOrElse(List())
+              .flatMap(r => Try(r.get(unorderedFieldKeys(fullName))).toOption.map(k => (k, r)))
+              .toMap
+            val r = y.flatMap(r =>
+              Option(r.get(name).asInstanceOf[java.util.List[GenericRecord]].asScala.toList))
+              .getOrElse(List())
+              .flatMap(r => Try(r.get(unorderedFieldKeys(fullName))).toOption.map(k => (k, r)))
+              .toMap
+            (l.keySet ++ r.keySet).flatMap(k => diff(l.get(k), r.get(k), fullName)).toList
```
Comment (danielblazevski) [inline ratatool-diffy/src/main/scala/com/spotify/ratatool/diffy/AvroDiffy.scala]: Seems like we're saying the diff of two records `x` and `y` that are like `Array[NestedRecord]` is computing by taking any exact matches of some subfield `x.a` and `y.a` of `NestedRecord` and comparing?  

I worry this could not be what some users would expect - e.g. if the arrays are equal size and all subfields are numeric but all slightly different, there would be no way to compute the diff.

Comment (danielblazevski) [reply ratatool-diffy/src/main/scala/com/spotify/ratatool/diffy/AvroDiffy.scala]: Synced offline.  

We agreed this does change the behavior since we no longer sort on `unorderedFieldKeys` in a way that'll mark all records as "unkonwn delta" if we want to compare on numeric fields that could be different.  But all use cases we know of are for strings from what I was told.  

Docs should be updated though, esp to clarify that `unorderedFieldKeys` is no longer a sorting key as the docs c currently say.

Comment (danielblazevski) [reply ratatool-diffy/src/main/scala/com/spotify/ratatool/diffy/AvroDiffy.scala]: Might also be good to write a test w/ this new behavior.  

The test would also aid as documentation for expected behavior.

**Comment by danielblazevski** [comment]
LGTM after modifying docs clarifying `unorderedFieldKeys` will no longer sort off that key + adding a test verifying we pick up the right records to diff.

File: ratatool-diffy/src/main/scala/com/spotify/ratatool/diffy/AvroDiffy.scala [hunk 6]
```diff
@@ -20,67 +20,79 @@ package com.spotify.ratatool.diffy
...
-  // scalastyle:off cyclomatic.complexity
-  private def diff(x: GenericRecord, y: GenericRecord, root: String): Seq[Delta] = {
-    def getField(f: String)(x: GenericRecord): AnyRef = {
-      x.get(f)
+  // scalastyle:off cyclomatic.complexity method.length
+  private def diff(x: Option[GenericRecord], y: Option[GenericRecord], root: String)
+  : Seq[Delta] = {
+    val schemaFields = (x, y) match {
+      case (Some(xVal), None) => xVal.getSchema.getFields.asScala.toList
+      case (_, Some(yVal)) => yVal.getSchema.getFields.asScala.toList
```
[the remark this answers is not in the document]
Comment (idreeskhan) [reply ratatool-diffy/src/main/scala/com/spotify/ratatool/diffy/AvroDiffy.scala]: I'll add a comment. It's because we assume LHS is backwards compatible with RHS, therefore RHS should have all necessary fields

File: ratatool-diffy/src/main/scala/com/spotify/ratatool/diffy/AvroDiffy.scala [hunk 7]
```diff
@@ -20,67 +20,79 @@ package com.spotify.ratatool.diffy
...
-              && unordered.exists(_.startsWith(s"$fullName."))
-              && unorderedFieldKeys.contains(fullName)) {
-            val a = sortList(x.get(name).asInstanceOf[java.util.List[GenericRecord]],
-              unorderedFieldKeys.get(fullName).map(getField))
-            val b = sortList(y.get(name).asInstanceOf[java.util.List[GenericRecord]],
-              unorderedFieldKeys.get(fullName).map(getField))
-            a.asScala.zip(b.asScala).flatMap{case (l, r) =>
-              diff(l.asInstanceOf[GenericRecord], r.asInstanceOf[GenericRecord], fullName)
-            }.toList
+            && unordered.contains(fullName)
```
[the remark this answers is not in the document]
Comment (idreeskhan) [reply ratatool-diffy/src/main/scala/com/spotify/ratatool/diffy/AvroDiffy.scala]: Thanks, nice catch

File: ratatool-diffy/src/main/scala/com/spotify/ratatool/diffy/AvroDiffy.scala [hunk 8]
```diff
@@ -20,67 +20,79 @@ package com.spotify.ratatool.diffy
...
 
-  // scalastyle:off cyclomatic.complexity
-  private def diff(x: GenericRecord, y: GenericRecord, root: String): Seq[Delta] = {
-    def getField(f: String)(x: GenericRecord): AnyRef = {
-      x.get(f)
+  // scalastyle:off cyclomatic.complexity method.length
+  private def diff(x: Option[GenericRecord], y: Option[GenericRecord], root: String)
+  : Seq[Delta] = {
+    val schemaFields = (x, y) match {
+      case (Some(xVal), None) => xVal.getSchema.getFields.asScala.toList
```
[the remark this answers is not in the document]
Comment (idreeskhan) [reply ratatool-diffy/src/main/scala/com/spotify/ratatool/diffy/AvroDiffy.scala]: asScala has laziness which causes weirdness and hard to parse errors. Easier to just avoid it imo

File: ratatool-diffy/src/main/scala/com/spotify/ratatool/diffy/AvroDiffy.scala [hunk 9]
```diff
@@ -20,67 +20,79 @@ package com.spotify.ratatool.diffy
...
-              unorderedFieldKeys.get(fullName).map(getField))
-            val b = sortList(y.get(name).asInstanceOf[java.util.List[GenericRecord]],
-              unorderedFieldKeys.get(fullName).map(getField))
-            a.asScala.zip(b.asScala).flatMap{case (l, r) =>
-              diff(l.asInstanceOf[GenericRecord], r.asInstanceOf[GenericRecord], fullName)
-            }.toList
+            && unordered.contains(fullName)
+            && unorderedFieldKeys.contains(fullName)) {
+            val l = x.flatMap(r =>
+              Option(r.get(name).asInstanceOf[java.util.List[GenericRecord]].asScala.toList))
```
[the remark this answers is not in the document]
Comment (idreeskhan) [reply ratatool-diffy/src/main/scala/com/spotify/ratatool/diffy/AvroDiffy.scala]: Hmm, I think the asScala won't immediately NPE but the toList will. I'll adjust

[the remark this answers is not in the document]
Comment (idreeskhan) [reply ratatool-diffy/src/main/scala/com/spotify/ratatool/diffy/AvroDiffy.scala]: Actually we will already have a properly wrapped option at this point so it's not a concern

File: ratatool-diffy/src/main/scala/com/spotify/ratatool/diffy/AvroDiffy.scala [hunk 10]
```diff
@@ -20,67 +20,79 @@ package com.spotify.ratatool.diffy
...
+              .getOrElse(List())
+              .flatMap(r => Try(r.get(unorderedFieldKeys(fullName))).toOption.map(k => (k, r)))
```
[the remark this answers is not in the document]
Comment (idreeskhan) [reply ratatool-diffy/src/main/scala/com/spotify/ratatool/diffy/AvroDiffy.scala]: Not sure it warrants a comment but I'll think of a way to make this more clear in the code

File: ratatool-diffy/src/main/scala/com/spotify/ratatool/diffy/AvroDiffy.scala [hunk 11]
```diff
@@ -20,67 +20,79 @@ package com.spotify.ratatool.diffy
...
+              .toMap
+            (l.keySet ++ r.keySet).flatMap(k => diff(l.get(k), r.get(k), fullName)).toList
           }
           else {
-            val a = sortList(x.get(name).asInstanceOf[java.util.List[GenericRecord]])
-            val b = sortList(y.get(name).asInstanceOf[java.util.List[GenericRecord]])
-            if (a == b) Nil else Seq(Delta(fullName, Option(a), Option(b), delta(a, b)))
+            val a = x.flatMap(r => Option(r.get(name).asInstanceOf[java.util.List[GenericRecord]]))
```
[the remark this answers is not in the document]
Comment (idreeskhan) [reply ratatool-diffy/src/main/scala/com/spotify/ratatool/diffy/AvroDiffy.scala]: Can you clarify what you mean, not sure I understand this
